pkg/adaptor/cloud/cloud.go

```diff
@@ -0,0 +1,420 @@
+}
+
+type sandboxID string
+
+type sandbox struct {
+	id               sandboxID
+	pod              string
+	namespace        string
+	instanceName     string // set by StartVM
+	instanceID       string // set by StartVM
+	netNSPath        string
+	agentProxy       proxy.AgentProxy
+	podNetworkConfig *tunneler.Config
+	cloudConfig      *cloudinit.CloudConfig
+}
+
+func (s *cloudService) addSandbox(sid sandboxID, sandbox *sandbox) error {
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if _, exists := s.sandboxes[sid]; exists {
+		return fmt.Errorf("sandbox %s already exists", sid)
+	}
+
+	s.sandboxes[sid] = sandbox
+
+	return nil
+}
+
+func (s *cloudService) getSandbox(id sandboxID) (*sandbox, error) {
+
+	sid := sandboxID(id)
+
+	if id == "" {
+		return nil, errors.New("empty sandbox id")
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	if _, exists := s.sandboxes[sid]; !exists {
+		return nil, fmt.Errorf("sandbox %s does not exist", sid)
+	}
+	return s.sandboxes[sid], nil
+}
+
+func (s *cloudService) removeSandbox(id sandboxID) error {
+	sid := sandboxID(id)
+	if id == "" {
+		return errors.New("empty sandbox id")
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	delete(s.sandboxes, sid)
+	return nil
+}
+
+func NewService(provider Provider, proxyFactory proxy.Factory, workerNode podnetwork.WorkerNode, podsDir, daemonPort string) Service {
+
+	s := &cloudService{
+		provider:     provider,
+		proxyFactory: proxyFactory,
+		sandboxes:    map[sandboxID]*sandbox{},
+		podsDir:      podsDir,
+		daemonPort:   daemonPort,
+		workerNode:   workerNode,
+	}
+	s.cond = sync.NewCond(&s.mutex)
+
+	return s
+}
+
+func (s *cloudService) setInstance(sid sandboxID, instanceID, instanceName string) error {
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	sandbox, ok := s.sandboxes[sid]
+	if !ok {
+		return fmt.Errorf("sandbox %s does not exist", sid)
+	}
+
+	sandbox.instanceID = instanceID
+	sandbox.instanceName = instanceID
+
+	s.cond.Broadcast()
+
+	return nil
+}
+
+func (s *cloudService) GetInstanceID(ctx context.Context, podNamespace, podName string, wait bool) (string, error) {
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	for {
+		for _, sandbox := range s.sandboxes {
+			if sandbox.namespace == podNamespace && sandbox.pod == podName {
+				return sandbox.instanceID, nil
+			}
+		}
+
+		if !wait {
+			return "", nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("failed to get instance ID: %w", ctx.Err())
+		default:
+		}
+
+		s.cond.Wait()
+	}
+}
+
+func (s *cloudService) Version(ctx context.Context, req *pb.VersionRequest) (*pb.VersionResponse, error) {
+	return &pb.VersionResponse{Version: Version}, nil
+}
+
+func (s *cloudService) CreateVM(ctx context.Context, req *pb.CreateVMRequest) (res *pb.CreateVMResponse, err error) {
+
+	defer func() {
+		if err != nil {
+			logger.Print(err)
+		}
+	}()
+
+	sid := sandboxID(req.Id)
+
+	if sid == "" {
+		err := errors.New("empty sandbox id")
+		logger.Print(err)
+		return nil, err
```

Comment:
nit: may be having a consistent error reporting will be helpful. Eg using `return nil, fmt.Errorf("empty sandbox id")`

Reply:
done